Add an after-define hook to field extensions

rmosolgo opened the pull request:
This would allow field extensions to define arguments on fields _without_ overriding something from the field's definition block. 

cc @cocoahero 

TODO: 

- [x] Name 
- [x] Docs
- [x] Remove breaking change to ConnectionExtension

cocoahero commented:
I think this is a good solution. My vote for naming would be `after_define`.

It does get a bit tedious to check for each arguments presence before applying the change, but I think for this particular edge case its fine. Perhaps a future improvement would be adding a class level `default_argument` DSL that would behind the scenes automatically apply the arguments in the `after_define` callback while checking for collisions. Example:

```ruby
class MyExtension < GraphQL::Schema::FieldExtension
  default_argument :first, Integer, required: false

  def after_define
    # this would happen in the super impl
    all_keywords = field.all_argument_definitions.map(&:keyword)
    all_keywords.uniq!
    self.default_argument_defs.each do |arg_def|
      next if all_keywords.include?(arg.keyword)
      field.add_argument(arg_def)
    end
  end
```

swalkinshaw commented:
Exposing the original arguments and the _added_ arguments would be useful overall. We already have a low-level extension which has to do a similar thing manually:

```ruby
Class MyExtension
  def apply
    argument_names = field.arguments.keys
    # some field.add_argument calls
    added_arguments = field.arguments.keys - argument_names

    options[:added_arguments] = added_arguments.map do |key|
      key.underscore.to_sym
    end
  end

  def resolve(object:, arguments:, context:)
    next_args = arguments.dup
    arguments = arguments.except(*options[:added_arguments])

    yield(object, arguments, next_args)
  end

  def after_resolve(value:, memo:, context:, **_args)
    arguments = memo
  end
end
```

rmosolgo commented:
Thanks for taking a look. I: 

- renamed it `after_define`
- added `.default_argument` and `#added_default_argument` for a workflow like the one in ConnectionExtension 
- Made ConnectionExtension only delete arguments that it added by default -- this is actually a breaking change, I should revert it. But I wanted to make sure it worked.

What do you think of that?

cocoahero commented:
This feels pretty good to me. I'll play with this branch in our core codebase today to see if its unblocks our upgrade.

cocoahero commented:
I noticed a slight issue with the current implementation where extensions added to the field after the definition block would not have their `after_define_apply` method called, and therefore default arguments and the `after_define` hook would not be used. For example:

```ruby
def custom_field_def_helper
  field_def = field(name, return_type, &block)
  field_def.extension(...) # This extension's `apply` would be called, but not `after_define_apply`.
end
```

This may be an unsupported scenario, but something that came up with our current usage. Completely possible that its a foot-gun situation 😅

rmosolgo commented:
It certainly _used_ to work, and I agree it _should_. But I don't know a better approach than to call both `apply` and `after_define` right away in that case.

rmosolgo commented:
Updated along those lines in a242d60d4 -- let me know if you see any red flags, otherwise I'm hoping to 🚢  1.13.1 this afternoon.
